[NET-1304] Security implications and documentation around key revocation grace period (#74)
ceramicnetwork/js-did: README.md

const didWithCap2 = new DID({provider, resolver: KeyResolver.getResolver(), capability: cacao})
```

## Security Considerations

> Ceramic allows for keys attached to DIDs to continue making updates to streams after revocation for a certain grace period. This is done to ensure that if multiple updates were committed to different nodes before the key was revoked, but the nodes chose to anchor commits at different times, the updates still make it through.

Reviewer: huh, this just made me realize.  Did we implement the key revocation grace period in a way that applies to all DIDs, not just 3IDs?  That seems like a mistake, the issues with key revocations and anchor timings are pretty specific to 3IDs.  @oed

Reviewer 2: I would say it applies to all DIDs since things signed right before the revocation might not be anchored yet. This is also relevant for CACAO, which I just realized doesn't use the revocation timeout yet:  https://github.com/ceramicnetwork/js-did/blob/main/src/did.ts#L334

> Ceramic allows for keys attached to DIDs to continue making updates to streams after revocation for a certain grace period. This is done to ensure that if multiple updates were committed to different nodes before the key was revoked, but the nodes chose to anchor commits at different times, the updates still make it through.

Reviewer: ```suggestion
Ceramic allows for keys attached to DIDs to continue making updates to streams after revocation for a certain grace period. This is done to avoid incorrectly rejecting valid signatures as being performed with a revoked key when in fact the key was valid at the time of the signature.  This can happen due to the implementation details of how Ceramic determines when a DID issued a signature as well as how the 3ID DID method built on Ceramic determines when an authentication key for a 3ID is revoked.  Since we cannot trust the system clock time on the client's machine when performing a signature or key revocation, Ceramic periodically "anchors" updates on a blockchain to get a timestamp before which we know the write must have happened.  Consider the following scenario:
```

Author: thank you for these suggestions, very helpful. i need to get better at writing these kind of documents.

Reviewer 2: I don't think we need to explicitly mention 3ID here.

1. Commit for an update is made to Node A
2. Key revocation commit is made to Node B

> 3. Node A later tries to anchor the commit made earlier

Reviewer: ```suggestion
3. Node B's key revocation commit is anchored before node A's update
```

> In this case, an external outlook might lead you to believe Node A's commit is invalid because the key had been revoked, but that's not necessarily true as in this case the commit was made earlier, just not anchored at that time. For this reason, a grace period is provided for updates created by revoked keys to still be verified successfully.

Reviewer: ```suggestion
In this case, an external outlook might lead you to believe Node A's commit is invalid because the key had been revoked when it was made, but that's not actually true as in this case the commit was in fact made earlier in real world time, it just wasn't assigned a timestamp by the anchoring system until after. For this reason, a grace period is provided for updates created by revoked keys to still be verified successfully.
```

> Specifically, the `verifyJWS` function in `dids` uses `options.revocationPhaseOutSecs` as the seconds representing the grace period within which signatures authored by a given key will still be considered valid.

Reviewer: ```suggestion
Specifically, the `verifyJWS` function in `dids` uses `options.revocationPhaseOutSecs` as the seconds representing the grace period within which signatures authored by a given key that otherwise appears to be revoked will still be considered valid.
```

> While this approach has it's use cases, it also implies that an attacker could potentially author signatures from a stolen key and commit updates to streams using that stolen key. Even if the original owner revoked it, the attacker has a grace period within which they can make fake commits.

Reviewer: ```suggestion
While this approach has it's use cases, it also implies that an attacker could potentially author signatures from a stolen key and commit updates to streams using that stolen key. Even if the original owner revoked it, the attacker has a grace period within which they can make fake commits.
```
```suggestion
While this approach helps solve the problem of incorrectly rejecting valid updates, it also implies that an attacker could potentially author signatures from a stolen key and commit updates to streams using that stolen key. Even if the original owner revoked the compromised key, the attacker has a grace period within which they can continue to make new commits.
```

Reviewer 2: Worth noting here is that best practice is likely to rotate keys on a regular basis to mitigate stolen keys.
